Improvements to 2D depiction and alignment/RMSD calculation

ptosco opened the pull request:
- Refactored the `straightenDepiction` code which is now much simpler and more readable and supports a `minimizeRotation` parameter
- added C++, Python and JS tests for the new `minimizeRotation` parameter
- refactored tests to use `CalcRMS()` rather than an internal implementation to compute RMS deviations
- Removed duplicated code in `CalcRMS()` and `getBestRMS()` and made their APIs consistent with respect to supported parameters
IMPORTANT NOTE: for backwards compatibility I set the `CalcRMS()` default for the  new symmetrizeConjugatedTerminalGroups to false as this parameter was not originally supported. @greglandrum: I would be very much in favor of setting this to true instead if you agree, even though it might change results for existing scripts, as I think it is a much more sensible default.
- Improved documentation to clarify the difference between `CalcRMS()` and `getBestRMS()`
- Added unit tests for `CalcRMS()` as there was none previously
- Added tests for the additional `CalcRMS()` and `getBestRMS()` parameters
- Added a new `getBestAlignmentTransform()` function
- The CFFI function `set_2d_coords_aligned()` now returns the matching atoms similarly to the C++, Python and JS counterparts IMPORTANT NOTE: this required an API change for the additional `char **` parameter used to return the match. Existing code using `set_2d_coords_aligned()` will fail to compile and will require a last `NULL` parameter to be added to compile again
- Removed duplicated code between CFFI `set_2d_coords_aligned()` and JS `generate_aligned_coords()`
- Added `has_2d_coords()` to the CFFI library
- `generate_aligned_coords()` now supports JSON parameters and the previous versions are deprecated
- `set_2d_coords_aligned()` and `generate_aligned_coords()` both support an `alignOnly` parameter (which defaults to `false`). If set to `true`, rather than re-generating a fresh 2D layout around `templateMol`, the existing coordinates (if any) are simply aligned to the provided `templateMol`. If the molecule has no coordinates, a set of 2D coordinates is generated independently of `templateMol `and then aligned to the provided `templateMol`
- avoid that when `acceptFailure` is false `set_2d_coords_aligned()` and `generate_aligned_coords()` overwrite existing coordinates

greglandrum commented on Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp:
```diff
@@ -75,10 +77,18 @@ PyObject *generateRmsdTransPyTuple(double rmsd, RDGeom::Transform3D &trans) {
   PyObject *rmsdItem = PyFloat_FromDouble(rmsd);
   PyTuple_SetItem(resTup, 0, rmsdItem);
   PyTuple_SetItem(resTup, 1, PyArray_Return(res));
+  if (match) {
```
This code ends up returning the match as a tuple of 2-tuples. That's consistent with what the C++ code does, but the python versions of `GetSubstructMatch()` return a tuple of ints. I think that would be preferable here in order to maintain consistency within the Python code.
`Code/GraphMol/Wrap/substructmethods.h` has a `convertMatches()` function which you can use for this.

ptosco replied:
That's not feasible. Since by design a `map` argument can be passed which may include `(probeAtomID, refAtomId)` partial matches between `prbMol` and `refMol`, in order to enable matching a larger molecule to a smaller molecule and vice versa (and subset thereof), the returned match can be itself a subset of both `prbMol` and `refMol` indices. Therefore, indices in both `prbMol` and `refMol` are required to enable a unequivocal mapping.

greglandrum commented on Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp:
```diff
@@ -115,7 +125,45 @@ PyObject *getMolAlignTransform(const ROMol &prbMol, const ROMol &refMol,
+      nAtms = aMapVec.front().size();
+    }
+  }
+  RDNumeric::DoubleVector *wtsVec = translateDoubleSeq(weights);
```
How about making this a `unique_ptr` so that you don't have to worry about deallocation?

ptosco replied:
I am all in favor of that, but the `delete` was there because I copy/pasted that code from another function in the same file, which is peppered with `delete` statements. So I got replaced all of them with `unique_ptr`s.

greglandrum replied:
thanks for that!

greglandrum commented on Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp:
```diff
@@ -693,11 +794,12 @@ BOOST_PYTHON_MODULE(rdMolAlign) {
       (python::arg("prbMol"), python::arg("refMol"), python::arg("prbId") = -1,
        python::arg("refId") = -1, python::arg("map") = python::object(),
        python::arg("maxMatches") = 1000000,
+       python::arg("symmetrizeConjugatedTerminalGroups") = false,
```
I agree that this should be `true`, but the corresponding entry in the "backwards incompatible changes" part of the release notes should be updated.

greglandrum commented on Code/GraphMol/MolAlign/Wrap/testMolAlign.py:
```diff
@@ -422,7 +422,7 @@ def test15MultiConfs(self):
     for tm in tms[1:]:
       prbMol.AddConformer(tm.GetConformer(), True)
-    self.failUnlessEqual(prbMol.GetNumConformers(), 50)
+    self.assertEqual(prbMol.GetNumConformers(), 50)
```
thanks for updating all of these

greglandrum commented on Code/MinimalLib/jswrapper.cpp:
```diff
@@ -183,6 +183,53 @@ JSReaction *get_rxn_no_details(const std::string &input) {
   return get_rxn(input, std::string());
 }
 
+std::string generate_aligned_coords_deprecated(JSMol &self,
```
These deprecations should probably be mentioned in the corresponding section of the release notes

ptosco replied:
Done.

greglandrum commented on Code/MinimalLib/cffiwrapper.h:
```diff
@@ -122,13 +122,15 @@ RDKIT_RDKITCFFI_EXPORT short fragment_parent(char **pkl, size_t *pkl_sz,
 
 // coordinates
 RDKIT_RDKITCFFI_EXPORT void prefer_coordgen(short val);
+RDKIT_RDKITCFFI_EXPORT short has_2d_coords(char *mol_pkl, size_t mol_pkl_sz);
```
The actual implementation of this checks whether or not the molecule has a conformer, not 2D coords.
How about renaming this to `has_conformer`?

ptosco replied:
Good point, I copy pasted the `set_2d_coords` signature and left `2d` in by mistake. I renamed to `has_coords` for consistency with the `set_*` functions.

greglandrum commented on Code/MinimalLib/cffi_test.c:
```diff
@@ -489,6 +490,78 @@ void test_modifications() {
   printf("--------------------------\n");
 }
 
+float **_get_coord_array(char *mpkl, size_t mpkl_size) {
```
wow, writing this shows serious dedication to testing. Thanks! :-)

ptosco replied:
Writing C code is like driving a classic car: no assisted steering and traction control but, oh, how charming.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -840,85 +845,84 @@ void generateDepictionMatching3DStructure(RDKit::ROMol &mol,
-    auto d_theta = fmod(-theta, INCR_DEG);
-    if (fabs(d_theta) > HALF_INCR_DEG) {
-      d_theta -= DepictorLocal::copySign(INCR_DEG, d_theta, ALMOST_ZERO);
+    auto d_theta = fmod(-theta, 30.0);
```
It seems like it would have been better to keep the constexpr definition of `INCR_DEG` rather than having the constant value 30.0 (or the dependent 15.0) repeated multiple times. Why did you make that change?

ptosco replied:
You are probably right, even though I don't think it particularly clarifies that code. Done.

greglandrum reviewed (changes requested):
Basically looks good; I made a number of smaller suggestions.

The changes to External/AvalonTools which somehow leaked in here are redundant (since those are already on master), but they shouldn't cause any problems

ptosco commented on Code/GraphMol/Depictor/RDDepictor.h:
```diff
@@ -237,17 +237,24 @@ RDKIT_DEPICTOR_EXPORT void generateDepictionMatching3DStructure(
-//! \brief Rotate the 2D depiction such that the majority of bonds have a
-//! 30-degree angle with the X axis.
+//! \brief Rotate the 2D depiction such that the majority of bonds have an
+//! angle with the X axis multiple of 30 degrees.
```
Done.

ptosco commented:
@greglandrum I think I have addressed all your comments. I have merged with `upstream/master` so the Avalon changes should not show up anymore. Thanks for reviewing!

greglandrum reviewed (approved):
LGTM
